staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -0,0 +1,177 @@
+*/
+
+package cel
+
+import (
+	"fmt"
+
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/interpreter"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// Validator parallels the structure of schema.Structural and includes the compiled CEL programs
+// for the x-kubernetes-validations of each schema node.
+type Validator struct {
+	Items      *Validator
+	Properties map[string]Validator
+
+	AdditionalProperties *Validator
+
+	compiledRules []CompilationResult
+
+	// Program compilation is pre-checked at CRD creation/update time, so we don't expect compilation to fail
+	// they are recompiled and added to this type, and it does, it is an internal bug.
+	// But if somehow we get any compilation errors, we track them and then surface them as validation errors.
+	compilationErr error
+
+	// isRoot is true if this validator node is for the root of a custom resource
+	isRoot bool
+}
+
+// NewValidator returns compiles all the CEL programs defined in x-kubernetes-validations extensions
+// of the Structural schema and returns a custom resource validator.
+func NewValidator(s *schema.Structural) Validator {
+	return validator(s, true)
+}
+
+func validator(s *schema.Structural, isRoot bool) Validator {
+	if isRoot {
+		// 'apiVersion', 'kind', 'metadata.name' and 'metadata.generateName' are always accessible to validator rules
+		// even if not specified in the schema
+		s = model.WithTypeAndObjectMeta(s)
+	}
+	compiledRules, err := Compile(s, isRoot)
+	result := Validator{compiledRules: compiledRules, compilationErr: err, isRoot: isRoot}
+	if s.Items != nil {
+		compiledItem := validator(s.Items, false)
+		result.Items = &compiledItem
+	}
+	if len(s.Properties) > 0 {
+		result.Properties = make(map[string]Validator, len(s.Properties))
+		for k, prop := range s.Properties {
+			compiledProp := validator(&prop, false)
+			result.Properties[k] = compiledProp
+		}
+	}
+	if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+		compiledProp := validator(s.AdditionalProperties.Structural, false)
+		result.AdditionalProperties = &compiledProp
+	}
+
+	return result
+}
+
+// Validate validates all x-kubernetes-validations rules in Validator against obj and returns any errors.
+func (s *Validator) Validate(fldPath *field.Path, sts *schema.Structural, obj interface{}) field.ErrorList {
+	if s == nil || obj == nil {
+		return nil
+	}
+
+	errs := s.validateExpressions(fldPath, sts, obj)
+	switch obj := obj.(type) {
+	case []interface{}:
+		return append(errs, s.validateArray(fldPath, sts, obj)...)
+	case map[string]interface{}:
+		return append(errs, s.validateMap(fldPath, sts, obj)...)
+	}
+	return errs
+}
+
+func (s *Validator) validateExpressions(fldPath *field.Path, sts *schema.Structural, obj interface{}) (errs field.ErrorList) {
+	if s.isRoot {
+		sts = model.WithTypeAndObjectMeta(sts)
```

Comment:
don't we also miss that case for `XEmbeddedObject:true` ?

Reply:
or rather above in the `validator` recursion. I expected a case like `validate(..., s.XEmbeddedObject)`. Or maybe there is some other way to inject the implicitly existing name, generatedName ?

Reply:
We injected it here: https://github.com/kubernetes/kubernetes/pull/106051/files#diff-ac18f4f94716f3f6e6908ef6d1329577dbf0779637eccb480cd9386c2cd0c79aR58

I suppose we could consolidate the logic for consistency, either entirely in `SchemaDeclType` or in `validateExpressions`. `validateExpressions` would probably be cleaner. I'll move it.